include/boost/spirit/home/x3/support/expectation.hpp

```diff
@@ -0,0 +1,397 @@
+#if !defined(BOOST_SPIRIT_X3_SUPPORT_EXPECTATION_HPP)
+#define BOOST_SPIRIT_X3_SUPPORT_EXPECTATION_HPP
+
+#if !defined(BOOST_SPIRIT_X3_THROW_EXPECTATION_FAILURE)
+# define BOOST_SPIRIT_X3_THROW_EXPECTATION_FAILURE 1
+#endif
+
+#include <boost/config.hpp> // for BOOST_SYMBOL_VISIBLE, BOOST_ATTRIBUTE_NODISCARD
+#include <boost/core/ignore_unused.hpp>
+#include <boost/spirit/home/x3/support/unused.hpp>
+#include <boost/spirit/home/x3/support/context.hpp>
+
+// We utilize `x3::traits::build_optional<...>` for customization point
+// instead of directly wrapping `expectation_failure` with `boost::optional`.
+// This would make it possible for the user to eliminate the usages of
+// `boost::optional<T>`, and use `std::optional<T>` everywhere.
+//
+// Note that we are intentionally including this header regardless of
+// the value of BOOST_SPIRIT_X3_THROW_EXPECTATION_FAILURE, since the
+// helper types defined in non-throwing version might still be required
+// when the users benchmark their application just by switching the
+// macro while keeping their implementation unmodified.
+//
+// This will make it possible for the users to unconditionally
+// inject `x3::expectation_failure_optional<Iterator>` into their parser,
+// safely assuming that the value is no-op in throwing mode.
+#include <boost/spirit/home/x3/support/traits/optional_traits.hpp>
+
+// This is required for partial specialization of relevant helpers.
+// TODO: Add a macro to discard all #includes of <boost/optional.hpp>.
+//       (this is TODO because it requires changes in `optional_traits.hpp`.)
+#include <boost/optional.hpp>
+#include <optional>
+
+#if BOOST_SPIRIT_X3_THROW_EXPECTATION_FAILURE // throwing mode
+# define BOOST_SPIRIT_X3_EXPECTATION_FAILURE_API BOOST_SYMBOL_VISIBLE
+# define BOOST_SPIRIT_X3_EXPECTATION_FAILURE_BASE : std::runtime_error
+# include <boost/throw_exception.hpp>
+# include <stdexcept>
+
+#else // non-throwing mode
+# define BOOST_SPIRIT_X3_EXPECTATION_FAILURE_API
+# define BOOST_SPIRIT_X3_EXPECTATION_FAILURE_BASE
+#endif
+
+#include <string>
+#include <type_traits>
+
+
+namespace boost { namespace spirit { namespace x3
+{
+    struct expectation_failure_tag;
+
+    template <typename Iterator>
+    struct BOOST_SPIRIT_X3_EXPECTATION_FAILURE_API
+        expectation_failure BOOST_SPIRIT_X3_EXPECTATION_FAILURE_BASE
+    {
+    public:
+        expectation_failure(Iterator where, std::string const& which)
+        #if BOOST_SPIRIT_X3_THROW_EXPECTATION_FAILURE
+            : std::runtime_error("boost::spirit::x3::expectation_failure"),
+        #else
+            :
+        #endif
+            where_(where), which_(which)
+        {}
+
+        BOOST_ATTRIBUTE_NODISCARD
+        constexpr Iterator const& where() const noexcept { return where_; }
+
+        BOOST_ATTRIBUTE_NODISCARD
+        constexpr std::string const& which() const noexcept { return which_; }
+
+    private:
+        Iterator where_;
+        std::string which_;
+    };
```

Comment:
This would obviously lead to ODR violations. Be careful with functions too. Better to use separate namespaces to be safe to not create any ODR violations.

Reply:
While I understand the basics of ODR violation, I'm quite not sure what's the proper way to implement workaround in a production quality library code. Could you elaborate on how I should separate the namespaces?

Reply:
> While I understand the basics of ODR violation, I'm quite not sure what's the proper way to implement workaround in a production quality library code. Could you elaborate on how I should separate the namespaces?

Without sufficient context, I'm not sure what @Kojoley exactly is referring to here (also with functions). I'm guessing though that this can happen when both  branches of BOOST_SPIRIT_X3_THROW_EXPECTATION_FAILURE are potentially possible in different translation units. Could you please elaborate, @Kojoley ?

Reply:
Fixed in https://github.com/boostorg/spirit/pull/788/commits/d9b53fedb0722e52be650576c20ddaf806273298